File: static/code/sample/tethered/gdl-hifngrprnt-b.js
// In this sample:
// Detect a fingerprint and store it in the system
// Compare the detected fingerprint with the stored fingerprints

import pkg_serialusb from 'dlserialusb';
const {SerialUSB} = pkg_serialusb

import pkg_duelink from 'duelink';
const {DUELinkController} = pkg_duelink

let duelink = new DUELinkController(new SerialUSB());
await duelink.Connect();

// Sleep helper (ms)
function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

// Methods
async function Initialize() {
    let init = 1;

    console.log("Wait for initialization, setting baudrate....");
    while (init !== 0) {
        init = parseInt(await duelink.Engine.ExecuteCommand("Init()"));
        await sleep(1000);
    }
    console.log("Initialization done.");
}

async function GetSavedCount() {
    const ret = parseInt(await duelink.Engine.ExecuteCommand("GetSavedCnt()"));
    return ret;
}

async function DeleteModel(location) {
    await duelink.Engine.ExecuteCommand(`DelModel(${location})`);
}

async function TakeImg() {
    const ret = parseInt(await duelink.Engine.ExecuteCommand("TakeImg()"));
    return ret;
}

async function ConvertImg(id) {
    // id: 1 = first time
    //     2 = second time when confirming the finger
    const ret = parseInt(await duelink.Engine.ExecuteCommand(`ConvertImg(${id})`));
    return ret;
}

async function CreateModel() {
    const ret = parseInt(await duelink.Engine.ExecuteCommand("CreateModel()"));
    return ret;
}

async function StoreModel(location) {
    const ret = parseInt(await duelink.Engine.ExecuteCommand(`StoreModel(${location})`));
    return ret;
}

async function FastSearch() {
    const ret = parseInt(await duelink.Engine.ExecuteCommand("FastSearch()"));
    return ret;
}

async function SetColor(color, blink, count) {
    // color: 0 = off, 1 = red, 2 = blue, 3 = purple
    const b = blink ? 1 : 0; // 1 = blink, 0 = always on
    // count: blink count, only applied when blink is true, 0 = blink forever
    await duelink.Engine.ExecuteCommand(`SetLed(${color},${b},${count})`);
}

async function DetectFinger() {
    // Detect a finger on the surface, not matching a stored fingerprint
    const ret = await duelink.Digital.Read(4, 1);
    return !ret;
}

(async () => {
    let s = await GetSavedCount();

    if (s > 0) {
        console.log("Saved fingerprint detected, deleting....");
        await DeleteModel(1);
        s = await GetSavedCount();
        console.log(`After deletion, count: ${s}`);
    }

    while (s === 0) {
        await sleep(1000);
        console.log("No saved fingerprint found, need to add.");

        if (await DetectFinger()) { // Pin 4 is low: user finger detected on surface
            // Set LED to blink red
            await SetColor(1, true, 0);
            console.log("Taking image of your fingerprint.....");

            while ((await TakeImg()) !== 0) {
                console.log("Bad image! Try again");
                await sleep(1000);
            }

            console.log("Please wait while converting....");
            while ((await ConvertImg(1)) !== 0) {
                console.log("Bad conversion! Try again");
                await sleep(1000);
            }

            console.log("Converted. Remove your finger.");

            while (await DetectFinger()) { // Wait until user removes finger
                await sleep(1);
            }

            await sleep(100);
            console.log("Place the same fingerprint again, we need to take image twice.");

            while (!(await DetectFinger())) { // Wait until user places finger again
                await sleep(100);
            }

            while ((await TakeImg()) !== 0) { // 0 = FINGERPRINT_OK
                console.log("Bad image! Try again");
                await sleep(1000);
            }

            console.log("Please wait while converting....");

            while ((await ConvertImg(2)) !== 0) {
                console.log("Bad conversion! Try again");
                await sleep(1000);
            }

            console.log("Converted");

            if ((await CreateModel()) === 0) { // Create model
                if ((await StoreModel(1)) === 0) { // Store to location 1
                    s = await GetSavedCount();
                    if (s > 0) {
                        console.log("Your finger is saved. Remove your finger.");
                        await SetColor(2, false, 0); // Set LED blue, stay on
                    }
                }
            }
        }

        await sleep(100);
    }

    while (await DetectFinger()) {
        await sleep(500);
        console.log("Remove your finger!!!!");
    }

    let found = false;

    while (!found) {
        if (await DetectFinger()) { // Pin 4 is low: user finger detected on surface
            await SetColor(3, false, 0); // Set LED purple, stay on
            console.log("Finger detected. Taking image of your finger.....");

            while ((await TakeImg()) !== 0) { // 0 = FINGERPRINT_OK
                console.log("Bad image! Try again");
                await sleep(1000);
            }

            console.log("Please wait while converting....");

            while ((await ConvertImg(1)) !== 0) {
                console.log("Bad conversion! Try again");
                await sleep(1000);
            }

            console.log("Converted");

            if ((await FastSearch()) === 0) {
                console.log("Your finger FOUND in the database");
                await SetColor(2, false, 0); // Set LED blue, stay on
                found = true;
            } else {
                console.log("Your finger does NOT match any in the database");
            }
        }

        await sleep(100);
    }
})();

File: static/code/sample/tethered/gdl-hifngrprnt-b.cs
// In this sample:
// Detect a fingerprint and store it in the system
// Compare the detected fingerprint with the stored fingerprints

using System;
using GHIElectronics.DUELink;

var availablePort = DUELinkController.GetConnectionPort();
var duelink = new DUELinkController(availablePort);

void Initialize() {
    var init = 1;

    Console.WriteLine("Wait for initialize, setting baudrate....");
    while (init != 0) {
        init = (int)duelink.Engine.ExecuteCommand("Init()");
        Thread.Sleep(1000);
    }
    Console.WriteLine("Initialize done.");
}

int GetSavedCount() {
    var ret = (int)duelink.Engine.ExecuteCommand("GetSavedCnt()");
    return ret;
}

void DeleteModel(int location) {
    duelink.Engine.ExecuteCommand($"DelModel({location})");
}

int TakeImg() {
    var ret = (int)duelink.Engine.ExecuteCommand("TakeImg()");
    return ret;
}

int ConvertImg(int id) {
    // id: 1: first time
    //     2: second time when confirm the finger
    var ret = (int)duelink.Engine.ExecuteCommand($"ConvertImg({id})");
    return ret;
}

int CreateModel() {
    var ret = (int)duelink.Engine.ExecuteCommand("CreateModel()");
    return ret;
}

int StoreModel(int location) {
    var ret = (int)duelink.Engine.ExecuteCommand($"StoreModel({location})");
    return ret;
}

int FastSearch() {
    var ret = (int)duelink.Engine.ExecuteCommand("FastSearch()");
    return ret;
}

void SetColor(int color, bool blink, int count) {
    // color: 0: off, 1:red; 2: blue; 3: purple
    var b = blink ? 1 : 0; // 1 blink; 0: always on
    // count: blink count, only applied when blink is true, 0: blink forever

    duelink.Engine.ExecuteCommand($"SetLed({color},{b},{count})");
}

bool DetectFinger() {
    // This detect finger on surface, not matched finger
    var ret = duelink.Digital.Read(4, DUELinkController.InputType.PullUp);
    return !ret;
}

var s = GetSavedCount();

if (s > 0) {
    Console.WriteLine("Detected saved, deleting....");
    DeleteModel(1);
    s = GetSavedCount();
    Console.WriteLine($"After deleted, count: {s}");
}

while (s == 0) {
    Thread.Sleep(1000);
    Console.WriteLine("No saved fingerprint found, need to add");

    if (DetectFinger()) { //# pin 4 is low: detected user finger on surface
        //duelink.Engine.ExecuteCommand("SetLed(1,1,0)"); //# Setting: Blink Red
        SetColor(1, true, 0); //Setting: Blink Red

        Console.WriteLine("Taking image of your fingerprint.....");

        while (TakeImg() != 0) {
            Console.WriteLine("Bad image! Try again");
            Thread.Sleep(1000);
        }
        Console.WriteLine("Please wait for converting....");
        while (ConvertImg(1) != 0) {
            Console.WriteLine("Bad Convert image! Try again");
            Thread.Sleep(1000);
        }

        Console.WriteLine("Converted. Remove your finger");

        while (DetectFinger()) { //# Wait till user remove their finger
            Thread.Sleep(1);
        }

        Thread.Sleep(100);

        Console.WriteLine("Place your same fingerprint again, we need to take image twice");

        while (!DetectFinger()) { //# Wait till user place their finger again
            Thread.Sleep(100);
        }

        while (TakeImg() != 0) { //# 0 FINGERPRINT_OK 
            Console.WriteLine("Bad image! Try again");
            Thread.Sleep(1000);
        }

        Console.WriteLine("Please wait for converting....");

        while (ConvertImg(2) != 0) {
            Console.WriteLine("Bad Convert image! Try again");
            Thread.Sleep(1000);
        }

        Console.WriteLine("Converted");

        if (CreateModel() == 0) { //# create model
            if (StoreModel(1) == 0) { //# store to location 1
                s = GetSavedCount();

                if (s > 0) {
                    Console.WriteLine("Your finger is saved. Remove your finger");
                    //duelink.Engine.ExecuteCommand("SetLed(2,0,0)"); //# Set led Blue
                    SetColor(2, false, 0); //Set led Blue, stay on
                }
            }
        }

    }
    Thread.Sleep(100);
}


while (DetectFinger()) {
    Thread.Sleep(500);
    Console.WriteLine("Remove your finger!!!!");
}

var found = false;

while (!found) {
    if (DetectFinger()) { //# pin 4 is low: detected user finger on surface
        //duelink.Engine.ExecuteCommand("SetLed(3,0,0)"); //# Turn led to purple
        SetColor(3, false, 0); //Set led purple, stay on
        Console.WriteLine("Detect finger. Taking image of your finger.....");

        while (TakeImg() != 0) { //# 0 FINGERPRINT_OK 
            Console.WriteLine("Bad image! Try again");
            Thread.Sleep(1000);
        }

        Console.WriteLine("Please wait for converting....");

        while (ConvertImg(1) != 0) {
            Console.WriteLine("Bad Convert image! Try again");
            Thread.Sleep(1000);
        }

        Console.WriteLine("Converted");

        if (FastSearch() == 0) {
            Console.WriteLine("Your finger FOUND in our database");
            //duelink.Engine.ExecuteCommand("SetLed(2,0,0)"); //# Turn led to Blue
            SetColor(2, false, 0); //Set led Blue, stay on
            found = true;
        }
        else {
            Console.WriteLine("Your finger does NOT match in our database");
        }
    }
    Thread.Sleep(100);
}


